runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkDefaultExecutableStageContext.java

```diff
@@ -64,19 +71,60 @@ public void close() throws Exception {
     jobBundleFactory.close();
   }
 
-  enum ReferenceCountingFactory implements Factory {
-    REFERENCE_COUNTING;
+  private static class JobFactoryState {
+    private final AtomicInteger counter = new AtomicInteger(0);
+    private final List<ReferenceCountingFlinkExecutableStageContextFactory> factories =
+        new ArrayList<>();
+    private final int maxFactories;
 
-    private static final ReferenceCountingFlinkExecutableStageContextFactory actualFactory =
-        ReferenceCountingFlinkExecutableStageContextFactory.create(
-            FlinkDefaultExecutableStageContext::create);
+    private JobFactoryState(int maxFactories) {
+      if (maxFactories == 0) {
+        // Default to num_cores - 1 so that we leave some resources available for the java process
+        this.maxFactories = Math.max(Runtime.getRuntime().availableProcessors() - 1, 1);
+      } else {
+        this.maxFactories = maxFactories;
+      }
+    }
+
+    private synchronized FlinkExecutableStageContext.Factory getFactory() {
+      int count = counter.getAndIncrement();
+
+      if (count < maxFactories) {
+        factories.add(
+            ReferenceCountingFlinkExecutableStageContextFactory.create(
+                FlinkDefaultExecutableStageContext::create));
+      }
+
+      return factories.get(count % maxFactories);
```

Comment:
Can I suggest the following? I got confused with the modulo operation potentially generating invalid indices. Feel free to use this (or not if you think it is not easier to read).

```java
    private synchronized FlinkExecutableStageContext.Factory getFactory() {
      counter++;
      if (factories.size() < maxFactories) {
        ReferenceCountingFlinkExecutableStageContextFactory newContextFactory = 
            ReferenceCountingFlinkExecutableStageContextFactory.create(FlinkDefaultExecutableStageContext::create)
        factories.add(newContextFactory);
        return newContextFactory;
      }
      return factories.get(counter % maxFactories);
    }
```

Reply:
We could also eliminate the counter altogether but it might be useful to have for debugging.

Reply:
Sorry, what I meant was to replace the `counter` with an `index`, e.g. 

```
index %= maxFactories
```

The counter is not really needed, just the current index.

Reply:
I agree it's not obvious that `count % maxFactories` will always produce a valid index. I'll change it to your version, which I think is a bit clearer about the core invariants.